src/pyfmi/simulation/assimulo_interface_fmi3.pyx

```diff
@@ -0,0 +1,611 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# distutils: define_macros=NPY_NO_DEPRECATED_API=NPY_1_7_API_VERSION
+
+"""
+This file contains code for mapping FMI3 FMUs to the Problem specifications
+required by Assimulo.
+"""
+
+import numpy as np
+cimport numpy as np
+
+import time
+import logging as logging_module
+from operator import index
+import scipy.sparse as sps
+from timeit import default_timer as timer
+
+cimport pyfmi.fmil_import as FMIL
+cimport pyfmi.fmi3 as FMI3
+from pyfmi.exceptions import (
+    FMUException,
+    InvalidOptionException,
+    FMIModel_Exception,
+    FMIModelException
+)
+
+try:
+    import assimulo
+    assimulo_present = True
+except Exception:
+    logging_module.warning(
+        'Could not load Assimulo module. Check pyfmi.check_packages()')
+    assimulo_present = False
+
+if assimulo_present:
+    from assimulo.problem cimport cExplicit_Problem
+    from assimulo.exception import AssimuloRecoverableError, TerminateSimulation
+
+cdef class FMIODE3(cExplicit_Problem):
+    """
+    An Assimulo Explicit Model extended to FMI3 interface.
+    """
+
+    def __init__(self, model, input = None, result_file_name = "",
+                 with_jacobian = False, start_time = 0.0, logging = False,
+                 result_handler = None, extra_equations = None, synchronize_simulation = False,
+                 number_of_diagnostics_variables = 0):
+        """
+        Initialize the problem.
+        """
+        self._model = model
+        self._adapt_input(input)
+        self.input_names = []
+        self.timings = {"handle_result": 0.0}
+
+        if type(model) == FMI3.FMUModelME3: # if isinstance(model, FMUModelME3):
+            self.model_me3 = model
+            self.model_me3_instance = 1
+        else:
+            self.model_me3_instance = 0
+
+        # Set start time to the model
+        self._model.time = start_time
+
+        self.t0 = start_time
+        self.y0 = self._model.continuous_states
+        self.problem_name = self._model.get_name()
+
+        [f_nbr, g_nbr] = self._model.get_ode_sizes()
+
+        self._f_nbr = f_nbr
+        self._g_nbr = g_nbr
+        self._A = None
+
+        # Used by Assimulo
+        self.state_events_use = g_nbr > 0
+        self.time_events_use = True
+
+        # If there is no state in the model, add a dummy state der(y)=0
+        if f_nbr == 0:
+            self.y0 = np.array([0.0])
+
+        # Determine the result file name
+        if result_file_name == '':
+            self.result_file_name = model.get_name()+'_result.txt'
+        else:
+            self.result_file_name = result_file_name
```

Comment:
Is it possible to move this out of FMI2 and 3 and into `cExplicit_Problem`? Maybe we can just do something such as `self.set_result_file_name()`
and 
```
def set_result_file_name(self, result_file_name):
  if result_file_name == '':
    self.result_file_name = model.get_name() + '_result.txt'
  else:
    self.result_file_name = result_file_name
```

Reply:
`cExplicit_Problem` is in Assimulo; so no on that. 

One could add another intermediate class in PyFMI, but I'd actually prefer to not introduce extra dependencies between the different FMI versions.

Reply:
Oh, I missed that (just resolve)